utils/keyboardShortcuts.ts

```diff
@@ -0,0 +1,55 @@
+import { useEffect, useRef } from "react";
+
+const KEY_TO_MIDI = {
+  a: 60,
+  w: 61,
+  s: 62,
+  e: 63,
+  d: 64,
+  f: 65,
+  t: 66,
+  g: 67,
+  y: 68,
+  h: 69,
+  u: 70,
+  j: 71,
+  k: 72
+};
+
+const keyboardEventToOctaveOffset = (event: KeyboardEvent): number => {
+  return event.shiftKey ? 12 : event.ctrlKey ? -12 : 0;
```

Comment:
I will try to avoid this so that the code is easier to read. It's fine if its longer.

Reply:
I agree that the extra indirection might be overkill. I moved it inline and declare the `OCTAVE_SEMITONES` constant (`NUM_OF_SEMITONES_PER_OCTAVE` would be more descriptive, but can be too long)

Reply:
> I agree that the extra indirection might be overkill. I moved it inline and declare the OCTAVE_SEMITONES constant (NUM_OF_SEMITONES_PER_OCTAVE would be more descriptive, but can be too long)

I think it's fine. We can optimise at other places without compromising on the DX.

Can you comment what you are doing here? I think a lot of programmers are not familiar with music and I want to include as many comments as I can so that even someone who doesn't know music understands whats going on. I know this will require lot of comments but I am willing to do that so that everyone can understand whats happening in the codebase. WDYT?

Reply:
sure. I added inline comments.

Reply:
Thank you